### PR TITLE
Support DHT over IPv6

### PR DESCRIPTION
**Pull request by ssiloti**
Note this has seen limited real-world testing because IPv6 doesn't make is easy to bind multiple local addresses like IPv4 does.

### CODE REVIEW

File: include/libtorrent/kademlia/node_entry.hpp [hunk 1]
```diff
@@ -69,8 +69,7 @@ struct TORRENT_EXTRA_EXPORT node_entry
 
 	node_id id;
 
-	address_v4::bytes_type a;
-	boost::uint16_t p;
+	union_endpoint endpoint;
```
Comment (arvidn) [inline include/libtorrent/kademlia/node_entry.hpp]: I'm a little bit worried about this blowing up memory usage for the IPv4 routing table. Do you think it would make sense to make this type a template, with the address::bytes_type as a template parameter?

Comment (arvidn) [inline include/libtorrent/kademlia/node_entry.hpp]: I guess the node itself would have to be a template as well in that case... hm.

File: include/libtorrent/kademlia/observer.hpp [hunk 2]
```diff
@@ -146,19 +144,11 @@ struct observer : boost::noncopyable
 
 	node_id m_id;
 
-	TORRENT_UNION addr_t
```
Comment (arvidn) [inline include/libtorrent/kademlia/observer.hpp]: is this change necessary? it will make the observer base class larger, and it's not obvious if there's any gain

File: include/libtorrent/kademlia/routing_table.hpp [hunk 3]
```diff
@@ -73,6 +73,29 @@ struct routing_table_node
 	bucket_t live_nodes;
 };
 
+struct ip_set
```
Comment (arvidn) [inline include/libtorrent/kademlia/routing_table.hpp]: it would be nice to have a unit test for this. you could make the class TORRENT_EXTRA_EXPORT.

File: simulation/test_dht.cpp [hunk 4]
```diff
@@ -40,14 +40,38 @@ POSSIBILITY OF SUCH DAMAGE.
...
+	item = "mutable item";
+	seq = 1;
+	std::vector<char> v;
+	lt::bencode(std::back_inserter(v), item);
+	lt::dht::sign_mutable_item(std::make_pair(v.data(), v.size()), std::make_pair(salt.data(), salt.size())
+		, seq, pk.data(), sk.data(), sig.data());
+	g_put_count++;
+}
+
 struct network_config : network_setup_provider
```
Comment (arvidn) [inline simulation/test_dht.cpp]: this class seems to be driving a specific test now. perhaps network_config is not a suitable name anymore. or perhaps the test-specific logic should be factored out into a separate class or function (I'm hoping to add some DHT tests too, and ideally tests should be as small as possible to test a specific feature/property)

File: src/kademlia/dht_tracker.cpp [hunk 5]
```diff
@@ -192,6 +235,9 @@ namespace libtorrent { namespace dht
 		if (e || m_abort) return;
 
 		m_dht.tick();
+#if TORRENT_USE_IPV6
+		m_dht6.tick();
```
Comment (arvidn) [inline src/kademlia/dht_tracker.cpp]: what do you think about making m_dht an array of 1 or two elements instead? I'm not sure, but it looks like some of these calls could just be put in a loop instead of duplicated

File: src/kademlia/dht_tracker.cpp [hunk 6]
```diff
@@ -285,7 +370,11 @@ namespace libtorrent { namespace dht
 	void dht_tracker::get_item(sha1_hash const& target
 		, boost::function<void(item const&)> cb)
 	{
-		m_dht.get_item(target, boost::bind(&get_immutable_item_callback, _1, cb));
+		boost::shared_ptr<bool> item_posted = boost::make_shared<bool>(false);
```
Comment (arvidn) [inline src/kademlia/dht_tracker.cpp]: single byte heap allocations seem a bit shady (not very efficient). There's no other object this (and the other heap allocated int) could be coalesced with?

File: src/kademlia/dht_tracker.cpp [hunk 7]
```diff
@@ -416,40 +533,52 @@ namespace libtorrent { namespace dht
 		n->list().push_back(entry(node));
 	}
 
+	void save_nodes(entry& ret, node const& dht, std::string const& key)
```
Comment (arvidn) [inline src/kademlia/dht_tracker.cpp]: neat. free function! it would be nice to have a test for it too, since it should be simple

File: include/libtorrent/kademlia/node_entry.hpp (hunk 1 above)
Comment (ssiloti) [inline include/libtorrent/kademlia/node_entry.hpp]: Yeah it's unfortunate, but as you say, the alternative would be to make the entire node a template. With typical routing table sizes we're talking about around 10KB extra memory usage from this. That shouldn't blow anyone's memory budget unless they're trying to run on a microcontroller.

File: include/libtorrent/kademlia/observer.hpp (hunk 2 above)
Comment (ssiloti) [inline include/libtorrent/kademlia/observer.hpp]: Not necessary, just simplifying.

File: src/kademlia/node.cpp [hunk 8]
```diff
@@ -497,8 +508,20 @@ struct ping_observer : observer
...
 				std::copy(nodes, nodes + 20, id.begin());
 				nodes += 20;
-				algorithm()->get_node().m_table.heard_about(id
-					, detail::read_v4_endpoint<udp::endpoint>(nodes));
+#if TORRENT_USE_IPV6
+				if (at == ipv4)
+#endif
+				{
+					algorithm()->get_node().m_table.heard_about(id
+						, detail::read_v4_endpoint<udp::endpoint>(nodes));
```
Comment (arvidn) [inline src/kademlia/node.cpp]: how about just reading the endpoint into a generic udp::endpoint in the if-block, and then just have a single call to heard_about()?

File: src/kademlia/dht_tracker.cpp (hunk 6 above)
Comment (ssiloti) [inline src/kademlia/dht_tracker.cpp]: There's no existing object with the appropriate lifetime. I could create a pool allocator for them in the dht_tracker, but to me it seems like premature optimization since no existing use case for BEP44 is going to be generating large numbers of get/put requests.

File: src/kademlia/node.cpp [hunk 9]
```diff
@@ -497,8 +508,20 @@ struct ping_observer : observer
...
+				}
+#if TORRENT_USE_IPV6
```
Comment (arvidn) [inline src/kademlia/node.cpp]: if you put the ipv6 case at the top, you'll save an #if/#endif pair

File: src/kademlia/node.cpp [hunk 10]
```diff
@@ -888,9 +909,7 @@ void node::incoming_request(msg const& m, entry& e)
 		sha1_hash target(msg_keys[0].string_ptr());
 
 		// TODO: 2 find_node should write directly to the response entry
```
Comment (arvidn) [inline src/kademlia/node.cpp]: I imagine you can delete this TODO comment now too

File: src/kademlia/node.cpp [hunk 11]
```diff
@@ -1167,12 +1185,38 @@ void node::incoming_request(msg const& m, entry& e)
...
+	// entry based on the protocol the request came in with
+	if (want.type() != bdecode_node::list_t)
+	{
+		nodes_t n;
+		m_table.find_node(info_hash, n, 0);
+		write_nodes_entry(r[native_nodes_key()], n);
+		return;
+	}
+
+	for (int i = 0; i < want.list_size(); ++i)
```
Comment (arvidn) [inline src/kademlia/node.cpp]: maybe a high level description of what this code does would be helpful. It's not obvious to me how a node (this class) has a map, m_nodes, with pointers to other nodes. It feels like an odd pointer circle. I may be misunderstanding something, but wouldn't it make more sense to have the table/map of nodes in the parent object (dht_tracker)? Now they just have a duplicate map? or perhaps this is a reference assigned by the parent object

File: src/kademlia/routing_table.cpp [hunk 12]
```diff
@@ -57,15 +57,6 @@ POSSIBILITY OF SUCH DAMAGE.
 
 using boost::uint8_t;
 
-#if BOOST_VERSION <= 104700
-namespace boost {
-size_t hash_value(libtorrent::address_v4::bytes_type ip)
```
Comment (arvidn) [inline src/kademlia/routing_table.cpp]: I imagine this was used for the unordered set of IPs. does ip_set use an unordered_set keyed on union_address now? (I seem to recall you had two sets, which seems right, but isn't this still necessary then?)

File: src/kademlia/routing_table.cpp [hunk 13]
```diff
@@ -77,12 +68,43 @@ void erase_one(T& container, K const& key)
 	container.erase(i);
 }
 
-routing_table::routing_table(node_id const& id, int bucket_size
+void ip_set::insert(address addr)
+{
+	if (addr.is_v4())
+		m_ip4s.insert(addr.to_v4().to_bytes());
+#if TORRENT_USE_IPV6
```
Comment (arvidn) [inline src/kademlia/routing_table.cpp]: I like to re-order these blocks, so the if-statement can be excluded when built without IPv6 support

File: src/kademlia/routing_table.cpp [hunk 14]
```diff
@@ -558,6 +580,10 @@ routing_table::add_node_status_t routing_table::add_node_impl(node_entry e)
 	INVARIANT_CHECK;
 #endif
 
+	// don't add if the address isn't the right type
+	if (!native_address(e.addr()))
```
Comment (arvidn) [inline src/kademlia/routing_table.cpp]: I would imagine this is a violation of a precondition of this function. In which case I think you should have an assert too

File: src/kademlia/node.cpp (hunk 11 above)
Comment (ssiloti) [inline src/kademlia/node.cpp]: m_nodes is a reference to a std::map in dht_tracker which maps the "n4" and "n6" names to their respective dht::node objects.

File: src/kademlia/rpc_manager.cpp [hunk 15]
```diff
@@ -444,6 +417,12 @@ bool rpc_manager::invoke(entry& e, udp::endpoint target_addr
 	// places a 'ro' key in the top-level message dictionary and sets its value to 1.
 	if (m_settings.read_only) e["ro"] = 1;
 
+	node& n = o->algorithm()->get_node();
+	if (!n.native_address(o->target_addr()))
```
Comment (arvidn) [inline src/kademlia/rpc_manager.cpp]: ah, native_address() means is_native_address().. it took me a minute to make sense of this code

File: src/kademlia/traversal_algorithm.cpp [hunk 16]
```diff
@@ -600,7 +603,14 @@ void traversal_observer::reply(msg const& m)
 			node_id id;
 			std::copy(nodes, nodes + 20, id.begin());
 			nodes += 20;
-			algorithm()->traverse(id, read_v4_endpoint<udp::endpoint>(nodes));
+#if TORRENT_USE_IPV6
```
Comment (arvidn) [inline src/kademlia/traversal_algorithm.cpp]: it seems like this snippet could be simplified. by just doing the reading of the endpoint in the if-block and having a single call to algorithm()->traverse(). and to move the IPv6 case to the top (to only have one #if/#endif block)

File: src/session_impl.cpp [hunk 17]
```diff
@@ -6605,9 +6605,18 @@ namespace aux {
 		set_external_address(ip, source_dht, source);
 	}
 
-	address session_impl::external_address()
+	address session_impl::external_address(address_type at)
 	{
-		return m_external_ip.external_address(address_v4());
+#if TORRENT_USE_IPV6
```
Comment (arvidn) [inline src/session_impl.cpp]: this interleaving is hard to read I think. could you initialize a variable address, to either v4 or v6 and then do the call unconditionally? and maybe coalesce the #if/#endif blocks

File: test/test_dht.cpp [hunk 18]
```diff
@@ -672,7 +677,7 @@ TORRENT_TEST(dht)
 	// 50 downloaders and 50 seeds
 	for (int i = 0; i < 100; ++i)
 	{
-		source = udp::endpoint(rand_v4(), 6000);
+		source = udp::endpoint((*rand_addr)(), 6000);
```
Comment (arvidn) [inline test/test_dht.cpp]: you don't actually have to dereference the function pointer to call it. If you don't like that special case, you could also just make it a function reference (which probably is better anyway, since we don't need to be able to re-bind this variable to any other function anyway)

File: src/kademlia/routing_table.cpp (hunk 14 above)
Comment (ssiloti) [inline src/kademlia/routing_table.cpp]: We can get here via node_seen() when a node issues a bootstrap using the non-native IP version.

File: src/kademlia/dht_tracker.cpp (hunk 6 above)
Comment (arvidn) [inline src/kademlia/dht_tracker.cpp]: makes sense

File: src/kademlia/dht_tracker.cpp [hunk 19]
```diff
@@ -231,49 +285,80 @@ namespace libtorrent { namespace dht
...
+
+	bool get_mutable_item_callback(item& it, bool authoritative
+		, boost::shared_ptr<mutable_item_ctx> ctx
+		, boost::function<void(item const&, bool)> f)
 	{
 		// the reason to wrap here is to control the return value
 		// since it controls whether we re-put the content
 		TORRENT_ASSERT(it.is_mutable());
-		f(it, authoritative);
+		authoritative = authoritative && --ctx->active_traversals == 0;
```
Comment (arvidn) [inline src/kademlia/dht_tracker.cpp]: would you mind splitting the decrement into a separate statement?

File: src/kademlia/dht_tracker.cpp (hunk 5 above)
Comment (ssiloti) [inline src/kademlia/dht_tracker.cpp]: The problem is that dht::node is neither CopyConstructible nor DefaultConstructible so it cannot be initialized via aggregate initialization in C++03. This would require C++11.

**Comment by arvidn** [comment]
In the future, I'm hoping to make simplify and generalize the listen sockets (and make it more explicit and configurable exactly which interface to listen on). This would apply to UDP sockets as well. Right now, the abstraction for the UDP sockets is strange, where it makes both an IPv4 and an IPv6 socket look like one. If you have two IPv4 interfaces you're out of luck.

What I'm wondering is, would it make sense to move the abstraction down further to support *any* number of DHT nodes, each sitting on top of a single UDP socket bound to a single interface. The case you cover right now (and that the current UDP socket design restricts us to) it a single IPv4 and a single IPv6 socket.

I'm primarily interested in this because I think it may simplify the code, not necessarily that arbitrarily multi-homed machines is all that common.

For reference, I have an (incomplete) patch I'm working on here: https://github.com/arvidn/libtorrent/tree/listen-socket-cleanup

**Comment by ssiloti** [comment]
I think supporting multi-homed clients is the Right Thing to do, but I also don't think it will simplify this code. Mostly it just turns if/else blocks and doubled calls into dht::node with searching and iterating over a container. It also complicates the question of which node to reach into when a request is received with a `want` key for an address family other than the one the request came in from. Should we try to pick a node which is bound to same interface, or just go with the closest node bound to an address with the same scope? The latter is probably the way to go but either way it's more complex than just having one node per address family.

**Comment by ssiloti** [comment]
So where do we stand with this arvid? Do you want to wait for your listen socket cleanups to land or take these changes as-is then refactor them as part of the listen socket cleanup? I'm happy to take care of the refactoring either way.

**Comment by arvidn** [comment]
I don't think it's worth waiting. I can't really commit to complete that patch anytime soon really anyway. It's probably best to take an incremental approach anyway.

**Comment by ssiloti** [comment]
Regarding this failure:

`***** test_dht.cpp:422 "TEST_EQUAL_ERROR: items_num.size(): 3 expected: 4" *****`

This happens randomly because the outcome this is testing is dependent on the output of rand(). Specifically it depends on the number of items which get a `num_peers` value of 1. I've looked at announce_immutable_items and can't figure out what this code is supposed to be doing:

```
for (int i = 0; i < 1000; ++i)
	{
		for (int j = 0; j < num_items; ++j)
		{
			if ((i % items[j].num_peers) == 0) continue;
```

**Comment by arvidn** [comment]
is that the test for the "DHT scrape" feature? where the number of peers are reported as a bloom filter. This does add a random component to the test and I've also seen it fail occasionally. I think with a deterministic random function the test is reasonable (say, if lt::random() would be seeded with a constant at the beginning of the test). That's provided the test uses lt::random() pervasively though.

Fixing this is on my list, it just keeps getting pushed out for other more pressing tests. (In fact, the last week or so I've spent working on a proper DHT simulation test, with a simulated network of nodes, I'm hoping to feel more confident about a lot of DHT behavior once I finish it).

**Comment by arvidn** [comment]
I don't think I've seen it fail 3 times in a row though. so there may be something going on. I restarted the tests

**Comment by ssiloti** [comment]
It's not testing DHT scrape. All it does is put some immutable items to a node repeatedly then try to get them back out. The way the test is setup and run seems needlessly complicated, hence my confusion as to what the motivation was for writing it the way it is.

**Comment by gubatron** [comment]
@ssiloti sorry for my ignorance, if we had this merged today and deployed extensively, would this eliminate NAT and double NAT issues between IPv6 compatible peers? or IPv6 is still subject to NAT hell?

It'd be really cool if this would finally get rid of NAT issues and peers could fetch magnets freshly shared almost instantly with DHT IpV6 lookups.

**Comment by ssiloti** [comment]
IPv6 deployments typically do not use NAT. I've heard of corporate networks using NAT66 but I don't think that's common. NAT66 is not nearly as troublesome for P2P applications because it still allows hosts to accept incoming connections.

**Comment by LordNyriox** [comment]
@ssiloti I don't know what rules @arvidn follows when merging PRs, but do you think you can rebase this at some point?  

Thanks, 
~ Nyriox

**Comment by ssiloti** [comment]
@LordNyriox Yes I'll be rebasing this at some point. I figure arvid is busy preparing the 1.1 release so I was panning on rebasing after that is done.

**Comment by LordNyriox** [comment]
@ssiloti OK.  But what if @arvidn decides he wants to include this in v1.1, but can't because it's out of date?  That's my only concern about this right now.  

Of course, it's up to you.  :}  

Thanks, 
~ Nyriox

**Comment by LordNyriox** [comment]
@ssiloti My personal philosophy is that one should rebase as often as possible, particularly when dealing with all the last-minute fixes (which can break rebase ability in some cases) that all pile up in the flurry of development immediately prior to new releases.  

Of course, that's just my own two bits.  ;} 

~ Nyriox

**Comment by LordNyriox** [comment]
@ssiloti:  I'd recommend you rebase this sooner rather than later.  It might be a good feature for the 1.1 release. 

@arvidn:  If ssiloti rebases this, do you think you can merge this feature for the 1.1 release?  

Thanks, 
~ Nyriox

**Comment by arvidn** [comment]
@LordNyriox no, I do not want significant new features in 1.1, it would just push out it being stable further

**Comment by LordNyriox** [comment]
@arvidn:  Sorry, I am a bit clueless at times.  I thought of this more as an optimization than a feature in itself.  Just shows how clueless I am.  :}

**Comment by LordNyriox** [comment]
@arvidn:  While we're on the subject, do you have a planned (or hoped) release date for 1.1?  I know you probably hear that a lot but I was unable to find the information anywhere.  

Thanks, 
~ Nyriox

**Comment by arvidn** [comment]
There has been a lot more issues than I expected (and some quite serious, as the ones ssiloti discovered in the disk cache logic). Assuming no more serious issues are encountered, I would expect a few weeks from now. I mostly have time to work down the issue list on weekends.

You can help make it go faster by testing!

**Comment by LordNyriox** [comment]
@arvidn:  Well, sure—if I had the slightest clue how to operate a compiler.  As it is, I'd probably end up with enough false bugs from poor compilation to keep you busy for years… :/  

Thanks, though.  :}

**Comment by arvidn** [comment]
I noticed you rebased this. is it ready to be merged?

**Comment by LordNyriox** [comment]
@arvidn:  You might as well.  If anything turns out to need to be changed it can go in a separate PR.

**Comment by gubatron** [comment]
can't wait for this to be merged, this is I think, one of those patches that can trigger a whole lot of change on the internet. No more NAT hell and hopefully way higher success rates when finding/posting objects to the DHT.
